mtt/src/mtt.rs

```diff
@@ -0,0 +1,688 @@
+        }
+    }
+
+    #[allow(clippy::drop_non_drop)]
+    fn set_range_type(
+        &mut self,
+        addr: SupervisorPageAddr,
+        len: usize,
+        memory_type: MttMemoryType,
+        get_l1_page: &mut dyn FnMut() -> Option<Page<InternalClean>>,
+    ) -> Result<bool> {
+        let l2_dir = self.inner.lock();
+
+        let mut mtt_needs_invalidation = false;
+        let mut callback = |addr: SupervisorPageAddr, page_size: PageSize, l2_entry: MttL2Entry| {
+            let region_memory_type = MttL2Directory::get_memory_type(addr, &l2_entry);
+            if region_memory_type != memory_type {
+                mtt_needs_invalidation = true;
+                let target_page_size = Mtt::page_size_from_entry(&l2_entry);
+                if target_page_size == page_size {
+                    MttL2Directory::set_memory_type(addr, l2_entry, memory_type);
+                } else {
+                    // We are going from 1GB -> 64MB or 64MB -> L1, so
+                    // just split to the level below, and retry
+                    match l2_entry {
+                        MttL2Entry::NonConfidential1G(nc_1g) => nc_1g.split(),
+                        MttL2Entry::Confidential1G(conf_1g) => conf_1g.split(),
+                        MttL2Entry::Mixed2M(mixed_2m) => {
+                            if let Some(l1_page) = get_l1_page() {
+                                mixed_2m.split_to_l1_entry(l1_page)
+                            } else {
+                                return ControlFlow::Break(Err(InsufficientL1Pages));
+                            }
+                        }
+                        MttL2Entry::L1Entry(_) => unreachable!(),
+                    }
+                    // Retry same address with the new type
+                    return ControlFlow::Continue(false);
+                }
+            }
+
+            // Advance to the next address in the region
+            ControlFlow::Continue(true)
+        };
+        let mut mtt_range = MttRange::new(addr, len, l2_dir)?;
+        mtt_range.range_iterator(&mut callback)?;
+        drop(callback);
+        Ok(mtt_needs_invalidation)
+    }
+
+    /// Marks the region of length `len` starting with `addr` as confidential in
+    /// the Memory Tracking Table (MTT). On success, the function returns whether
+    /// the caller should make an ECALL to invalidate the MTT.
+    /// `len` must be a multiple of a 4K-page size. The region cannot exceed the
+    /// maximum physical page address (currently defined as 46-bits of PA).
+    /// `get_l1_page` will be called to allocate a 4K-page if the function determines
+    /// that the region mapping requires a MTT mapping at a 4K granularity.
+    /// The callback must ensure that the allocated page isn't aliased and remains allocated
+    /// until it's explictly released (TBD), and it must not perform any MTT related
+    /// operations as it will result in a deadlock.
+    pub fn set_confidential(
```

Comment:
see my comment on the previous revision about `set_confidential()` / `set_non_confidential()` being `unsafe`.